### PR TITLE
#3980 Support array of functions for onPrepare and onComplete hooks

### PR DESCRIPTION
**Pull request by naddison**
## Proposed changes

I have made changes required for onPrepare and onComplete to be treated like the other hooks. Your configuration file can support onPrepare and onComplete as a single function, or an array of functions. The ConfigParser defaults onPrepare and onComplete to arrays (like the other hooks), and pushes any new onPrepare and onComplete hooks from file config into the default arrays.

This is response to https://github.com/webdriverio/webdriverio/issues/3980

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by naddison** [comment]
> Code change looks good to me. I think we should add some documentation around failing the test process if your onComplete hooks fails.

@christian-bromann @mgrybyk 

The onComplete hook started failing the test run in this commit with the new logger. https://github.com/webdriverio/webdriverio/commit/270cca4bda82da90aacf53dead54ead45513a2c2 Did we intend for this and want to keep it? Or do we want to continue if the onComplete hook fails?

**Comment by mgrybyk** [comment]
It should not fail in onComplete hook but report failure after  because otherwise runner is not finishing properly that leads to some issues in different modes

**Comment by mgrybyk** [comment]
If you see a way to improve it somehow, feel free

File: packages/wdio-cli/src/utils.js [hunk 1]
```diff
@@ -17,6 +17,44 @@ export async function runServiceHook (launcher, hookName, ...args) {
     }
 }
 
+/**
+ * Run onPrepareHook in Launcher
+ * @param {Array|Function} onPrepareHook - can be array of functions or single function
+ * @param {Object} config
+ * @param {Object} capabilities
+ */
+export async function runOnPrepareHook(onPrepareHook, config, capabilities) {
```
Comment (christian-bromann) [inline packages/wdio-cli/src/utils.js]: One issue with this .. this is not backwards compatible. In order to not break every test we should ensure that functions are moved into an array.

File: docs/ConfigurationFile.md [hunk 2]
```diff
@@ -323,7 +323,8 @@ exports.config = {
     afterSession: function (config, capabilities, specs) {
     },
     /**
-     * Gets executed after all workers got shut down and the process is about to exit.
+     * Gets executed after all workers got shut down and the process is about to exit. An error
+     * thrown in the onComplete hook will result in the test run failing.
```
Comment (christian-bromann) [inline docs/ConfigurationFile.md]: Can we make this update everywhere?

File: packages/wdio-cli/src/utils.js (hunk 1 above)
Comment (naddison) [reply packages/wdio-cli/src/utils.js]: The onPrepare and onComplete functions are added to the root config array in the ConfigParser.addFileConfig() -> addService() function now.

File: docs/ConfigurationFile.md (hunk 2 above)
Comment (naddison) [reply docs/ConfigurationFile.md]: I'm not sure where else you want this updated, I can't find any other description of onComplete to update in any .md file.

File: packages/wdio-cli/src/utils.js (hunk 1 above)
Comment (naddison) [reply packages/wdio-cli/src/utils.js]: These should be backwards compatible regardless of the ConfigParser change.

File: docs/ConfigurationFile.md (hunk 2 above)
Comment (christian-bromann) [reply docs/ConfigurationFile.md]: e.g. /packages/wdio-cli/src/templates/wdio.conf.tpl.ejs .. there are a couple of wdio.conf.js examples in the repo

Comment (naddison) [reply docs/ConfigurationFile.md]: Ah okay, I found the culprits and updated them.

**Pull request merged**